graph/boruvka.cpp

```diff
@@ -0,0 +1,191 @@
+            component_size[x] += component_size[y];
+        }
+    }
+};
+
+/**
+ * @brief Boruvka's algorithm to compute the minimum spanning tree cost.
+ * @param num_nodes Number of nodes in the graph.
+ * @param num_edges Number of edges in the graph.
+ * @param edges Edges of the graph.
+ * @return Minimum spanning tree cost.
+ */
+long long boruvka(
+    int num_nodes, int num_edges,
+    const std::vector<std::pair<int, std::pair<int, int>>>& edges) {
+    union_find UF(num_nodes);
+
+    // In the beginning, each vertex represents an individual component
+    int total_components = num_nodes;
+
+    std::vector<int> cheapest_edge_from(num_nodes, -1);
+    long long total_cost = 0;
+
+    // This implementation assumes a connected graph as input. Algorithm can be
+    // adapted to find the minimum spanning forest by stopping whenever we
+    // iterate over all the edges and cannot find a new candidate edge to unite
+    // two components.
+    while (total_components > 1) {
+        // resetting the values because the components
+        fill(cheapest_edge_from.begin(), cheapest_edge_from.end(), -1);
+
+        // We now want to go over all edges and find the cheapest edge that
+        // leaves each one of the remaining components.
+        for (int e = 0; e < num_edges && total_components > 1; ++e) {
+            int from = edges[e].second.first;
+            int to = edges[e].second.second;
+            int weight = edges[e].first;
+
+            int from_component = UF.find(from);
+            int to_component = UF.find(to);
+
+            if (from_component == to_component) {
+                continue;
+            }
+
+            if (cheapest_edge_from[from_component] == -1 ||
+                edges[cheapest_edge_from[from_component]].first > weight) {
+                cheapest_edge_from[from_component] = e;
+            }
+
+            if (cheapest_edge_from[to_component] == -1 ||
+                edges[cheapest_edge_from[to_component]].first > weight) {
+                cheapest_edge_from[to_component] = e;
+            }
+
+            for (int i = 0; i < num_nodes; ++i) {
+                if (cheapest_edge_from[i] != -1) {
+                    int from = edges[cheapest_edge_from[i]].second.first;
+                    int to = edges[cheapest_edge_from[i]].second.second;
+                    int weight = edges[cheapest_edge_from[i]].first;
+
+                    if (UF.find(from) != UF.find(to)) {
+                        UF.unite(from, to);
+                        total_cost += weight;
+                        --total_components;
+                    }
+                }
+            }
+        }
+    }
+
+    return total_cost;
+}
+
+/**
+ * @brief Test function for Boruvka's algorithm.
+ */
+void tests() {
+    std::vector<std::pair<int, std::pair<int, int>>> edges;
+    edges.push_back({1, {0, 1}});
+    edges.push_back({2, {0, 2}});
+    edges.push_back({3, {1, 2}});
+    edges.push_back({4, {1, 3}});
```

Comment:
as mentioned above if you really want to use int then might want to add tests for cases when some of the values are negative